File: 2044_CountNumberOfMaximumBitwiseOrSubsets.py
from typing import List

class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        n = len(nums)
        
        maxOr = 0
        for num in nums:
            maxOr |= num

        memo = {}
        
        def dfs(index, currOr):
            if index==n:
                return int(currOr==maxOr)
            
            key = (index,currOr)
            if key in memo:
                return memo[key]

            memo[key] = dfs(index+1, currOr) + dfs(index+1, currOr|nums[index])
            return memo[key]

        return dfs(0,0)                


# Input: nums = [3,2,1,5]
# Output: 6

# 1 = [1,3,5]
# 10 = [2,3]
# 100 = [5]

# 

File: 2044_CountNumberOfMaximumBitwiseORSubsets.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
private:
    int n, maxOR = 0;
    unordered_map<int, int> memo;

    int dfs(const vector<int>& nums, int index, int currOR){
        if (index == n) return int(currOR == maxOR);

        int key = (currOR<<4) | index;
        if (memo.find(key)!=memo.end()) return memo[key];

        memo[key] = dfs(nums, index+1, currOR) + dfs(nums, index+1, currOR|nums[index]);

        return memo[key];

    }
public:
    int countMaxOrSubsets(vector<int>& nums) {
        for (auto& num: nums)   maxOR |= num;

        n = nums.size();

        return dfs(nums, 0, 0);
    }
};